File: MIDIJava/src/io/github/seanboyy/Midi/Midi.java
package io.github.seanboyy.Midi;

import java.util.ArrayList;

public class Midi {

    private ArrayList<Chunk> chunks;

    public Midi() {
        this.chunks = new ArrayList<>();
    }

    public Midi(ArrayList<Chunk> chunks) {
        this.chunks = new ArrayList<>();
        this.chunks.addAll(chunks);
    }

    @Override
    public String toString(){
        return Formatter.formatMidi(this);
    }

    public byte[] toBitString() {
        byte[] bitString = new byte[getLength()];
        int[] cursor = {0};
        for (Chunk chunk : chunks) {
            bitString = StreamingHelper.streamObjectToBitString(cursor, bitString, chunk, chunk.getLength());
        }
        return bitString;
    }

    void fromBitString(byte[] bitString, int fileSize) {
        int[] cursor = {0};
        chunks = new ArrayList<>();
        int signature = StreamingHelper.getIntFromBitString(cursor, bitString);
        assert (signature == MidiConstants.MIDI_SIGNATURE);
        int length = StreamingHelper.getIntFromBitString(cursor, bitString);
        short format = StreamingHelper.getShortFromBitString(cursor, bitString);
        short tracks = StreamingHelper.getShortFromBitString(cursor, bitString);
        short division = StreamingHelper.getShortFromBitString(cursor, bitString);
        chunks.add(new HeaderChunk(length, format, tracks, division));
        while (cursor[0] < fileSize) {
            signature = StreamingHelper.getIntFromBitString(cursor, bitString);
            assert (signature == MidiConstants.TRACK_SIGNATURE);
            length = StreamingHelper.getIntFromBitString(cursor, bitString);
            int endOfTrack = length + cursor[0];
            ArrayList<Event> events = new ArrayList<>();
            while (cursor[0] < endOfTrack) {
                VariableLengthValue deltaTime = StreamingHelper.getVariableLengthValueFromBitString(cursor, bitString);
                byte eventType = StreamingHelper.getByteFromBitString(cursor, bitString);
                if (eventType == MidiConstants.META_EVENT) {
                    byte metaEventType = StreamingHelper.getByteFromBitString(cursor, bitString);
                    if (metaEventType == MidiConstants.END_OF_TRACK)
                        assert (cursor[0] + 1 == endOfTrack);
                    VariableLengthValue eventLength = StreamingHelper.getVariableLengthValueFromBitString(cursor, bitString);
                    ArrayList<Byte> eventData = new ArrayList<>();
                    if (eventLength.toNumber() != 0) {
                        for (int i = 0; i < eventLength.toNumber(); ++i) {
                            eventData.add(StreamingHelper.getByteFromBitString(cursor, bitString));
                        }
                    }
                    events.add(new MetaEvent(deltaTime, metaEventType, eventLength, eventData));
                } else if (Integer.compareUnsigned(eventType, MidiConstants.META_EVENT) < 0 && Integer.compareUnsigned(eventType, MidiConstants.SYSEX_EVENT) >= 0) {
                    VariableLengthValue eventLength = StreamingHelper.getVariableLengthValueFromBitString(cursor, bitString);
                    ArrayList<Byte> eventData = new ArrayList<>();
                    if (eventLength.toNumber() != 0) {
                        for (int i = 0; i < eventLength.toNumber(); ++i) {
                            eventData.add(StreamingHelper.getByteFromBitString(cursor, bitString));
                        }
                    }
                    events.add(new SysexEvent(deltaTime, eventType, eventLength, eventData));
                } else if (Integer.compareUnsigned(eventType, MidiConstants.SYSEX_EVENT) < 0 && Integer.compareUnsigned(eventType, MidiConstants.NOTE_OFF) >= 0) {
                    ArrayList<Byte> eventData = new ArrayList<>();
                    if (Integer.compareUnsigned(eventType, MidiConstants.PROGRAM) < 0 || Integer.compareUnsigned(eventType, MidiConstants.PITCH) >= 0) {
                        eventData.add(StreamingHelper.getByteFromBitString(cursor, bitString));
                        eventData.add(StreamingHelper.getByteFromBitString(cursor, bitString));
                    } else {
                        eventData.add(StreamingHelper.getByteFromBitString(cursor, bitString));
                    }
                    events.add(new MidiEvent(deltaTime, eventType, eventData));
                }
            }
            chunks.add(new TrackChunk(length, events));
        }
    }

    public int getLength() {
        int length = 0;
        for (Chunk chunk : chunks) {
            length += chunk.getLength();
        }
        return length;
    }

    ArrayList<Chunk> getChunks(){
        return chunks;
    }
}

File: MIDIC#/MIDILibrary/MIDILibrary/Midi.cs
using System.Collections.Generic;

namespace MIDILibrary
{
    public class Midi
    {
        public List<IChunk> Chunks { get; }

        public Midi()
        {
            Chunks = new List<IChunk>();
        }

        public Midi(IEnumerable<IChunk> chunks)
        {
            Chunks = new List<IChunk>();
            Chunks.AddRange(chunks);
        }

        public byte[] ToBitString()
        {
            var bitString = new byte[GetLength()];
            uint cursor = 0;
            foreach (var chunk in Chunks)
            {
                StreamingHelper.StreamObjectToBitString(ref cursor, ref bitString, chunk, chunk.GetLength());
            }
            return bitString;
        }

        public void FromBitString(byte[] bitString, int fileSize)
        {
            uint cursor = 0;
            var signature = StreamingHelper.GetIntFromBitString(ref cursor, ref bitString);
            if (signature != MidiConstants.MIDI_SIGNATURE) throw new MidiException("Unexpected values in signature. Expected MThd (0x4D546864)");
            var length = StreamingHelper.GetIntFromBitString(ref cursor, ref bitString);
            var format = StreamingHelper.GetShortFromBitString(ref cursor, ref bitString);
            var tracks = StreamingHelper.GetShortFromBitString(ref cursor, ref bitString);
            var division = StreamingHelper.GetShortFromBitString(ref cursor, ref bitString);
            Chunks.Add(new HeaderChunk(length, format, tracks, division));
            while(cursor < fileSize)
            {
                signature = StreamingHelper.GetIntFromBitString(ref cursor, ref bitString);
                if (signature != MidiConstants.TRACK_SIGNATURE) throw new MidiException("Unexpected values in signature. Expected MTrk (0x4D54726B)");
                length = StreamingHelper.GetIntFromBitString(ref cursor, ref bitString);
                var endOfTrack = length + cursor;
                var events = new List<IEvent>();
                while(cursor < endOfTrack)
                {
                    var deltaTime = StreamingHelper.GetVariableLengthValueFromBitString(ref cursor, ref bitString);
                    var eventType = StreamingHelper.GetByteFromBitString(ref cursor, ref bitString);
                    if(eventType == MidiConstants.META_EVENT)
                    {
                        var metaEventType = StreamingHelper.GetByteFromBitString(ref cursor, ref bitString);
                        if (metaEventType == MidiConstants.END_OF_TRACK)
                            if (cursor + 1 != endOfTrack) throw new MidiException("Unexpected End of Track event");
                        var eventLength = StreamingHelper.GetVariableLengthValueFromBitString(ref cursor, ref bitString);
                        var eventData = new List<byte>();
                        if(eventLength.ToNumber() != 0)
                            for(uint i = 0; i < eventLength.ToNumber(); ++i)
                            {
                                eventData.Add(StreamingHelper.GetByteFromBitString(ref cursor, ref bitString));
                            }
                        events.Add(new MetaEvent(deltaTime, metaEventType, eventLength, eventData));
                    }
                    else if(eventType < MidiConstants.META_EVENT && eventType >= MidiConstants.SYSEX_EVENT)
                    {
                        var eventLength = StreamingHelper.GetVariableLengthValueFromBitString(ref cursor, ref bitString);
                        var eventData = new List<byte>();
                        if(eventLength.ToNumber() != 0)
                            for(uint i = 0; i < eventLength.ToNumber(); ++i)
                            {
                                eventData.Add(StreamingHelper.GetByteFromBitString(ref cursor, ref bitString));
                            }
                        events.Add(new SysexEvent(deltaTime, eventType, eventLength, eventData));
                    }
                    else if(eventType < MidiConstants.SYSEX_EVENT && eventType >= MidiConstants.NOTE_OFF)
                    {
                        var eventData = new List<byte>();
                        if(eventType < MidiConstants.PROGRAM || eventType >= MidiConstants.PITCH)
                        {
                            eventData.Add(StreamingHelper.GetByteFromBitString(ref cursor, ref bitString));
                            eventData.Add(StreamingHelper.GetByteFromBitString(ref cursor, ref bitString));
                        }
                        else
                            eventData.Add(StreamingHelper.GetByteFromBitString(ref cursor, ref bitString));
                        events.Add(new MidiEvent(deltaTime, eventType, eventData));
                    }
                }
                Chunks.Add(new TrackChunk(length, events));
            }
        }

        public uint GetLength()
        {
            uint length = 0;
            foreach(var chunk in Chunks)
            {
                length += chunk.GetLength();
            }
            return length;
        }
    }
}
